cli/parser/parser.go

```diff
@@ -355,41 +354,130 @@ type CommandLineSchema struct {
+			// isn't a startup option, but if it's an expansion option we need to emit
+			// a warning.
+			if command == "startup" {
+				// Unlike command options, startup options don't allow specifying
+				// values for options that do not require values.
+				return nil, false, fmt.Errorf("in option --%q: option %q does not take a value", optName, d.Name)
+			}
+			if !d.HasNegative {
+				// This is an expansion option with a specified value. Expansion options
+				// ignore values and output a warning. Since we canonicalize the options
+				// and remove the value ourselves, we should output the warning instead.
+				log.Warnf("option '%s' is an expansion option. It does not accept values, and does not change its expansion based on the value provided. Value '%s' will be ignored.", d.Name, v)
+				v = ""
 			}
 		}
-		// Check for starlark flags, which won't be listed in the schema that we
-		// parsed from "bazel help" output. All bazel commands support starlark
-		// flags like "--@repo//path:name=value". Even non-build commands like
-		// "bazel info" support these, but just ignore them.
-		if strings.Contains(opt, ":") || strings.Contains(opt, "/") {
-			return true
+		option := &Option{OptionDefinition: d, Value: v}
+		if d.HasNegative {
+			if b, err := option.AsBool(); err == nil {
+				// Normalize this boolean value
+				if b {
+					option.Value = "1"
+				} else {
+					option.Value = "0"
+				}
+			}
+		}
+		return option, d.RequiresValue && !hasValue, nil
+	}
+	if boolOptName, found := strings.CutPrefix(optName, "no"); found {
+		if d, ok := s.ByName[boolOptName]; ok && d.HasNegative {
+			if hasValue {
+				// This is a negative boolean value (of the form "--noNAME") with a
+				// specified value, which is unsupported.
+				return nil, false, fmt.Errorf("Unexpected value after boolean option: %s", optName)
+			}
+			return &Option{OptionDefinition: d, Value: "0"}, false, nil
 		}
-		return false
-	} else if strings.HasPrefix(opt, "-") {
-		// Short-form arg
-		opt = strings.TrimPrefix(opt, "-")
-		_, ok := s.CommandOptionDefinitions.ByShortName[opt]
-		return ok
 	}
-	return false
+	if command != "startup" {
+		// Check for starlark flags we haven't encountered yet, which won't be
+		// listed in the definitions we parsed from the flags collection proto. All
+		// bazel commands support starlark flags like "--@repo//path:name=value".
+		// Even non-build commands like "bazel info" support these, but just ignore
+		// them; however, they are not supported as startup flags. If it's a valid
+		// starlark flag, we add it to the set of option definitions in the parser
+		// in case we encounter it again.
+		d := s.parseStarlarkOptionDefinition(optName)
+		if d != nil {
+			// No need to check if this option already exists since we never reach
+			// this code if it does.
+			s.ForceAddOptionDefinition(d)
+			return &Option{OptionDefinition: d, Value: v}, false, nil
```

Comment:
A single logical Starlark option could be referenced in different ways (e.g. with and without a repo name for the main repo). We need to make sure that the relative order of all Starlark flags is preserved - is that the case?

Reply:
No, I didn't know that; I will revisit this, then. Though it also currently isn't the case that we handle them that way, so fixing that bug might be out of scope for this PR. Okay if I fix in a follow-up? This one is already bigger than I'd prefer.